File: crates/faf-domain/src/reducer.rs
//! The reducer: the entire mutation surface of the application.
//!
//! Pure, total, no IO, no async. Dispatches each [`AppEvent`] to the owning
//! slice reducer. To add a slice, add one match arm here (ARCHITECTURE.md §3.3).

use crate::state::{
    auth, chat, client_update, coop, galactic_war, install, leaderboard, lobby, map_generator,
    maps, mods, nav, notifications, player_card, replays, reporting, reviews, session, settings,
    social, tourney, tutorials, uploads,
};
use crate::{AppEvent, AppState};

pub fn reduce(state: &mut AppState, event: &AppEvent) {
    match event {
        AppEvent::Session(e) => session::reduce(&mut state.session, e),
        AppEvent::Auth(e) => auth::reduce(&mut state.auth, e),
        AppEvent::Nav(e) => nav::reduce(&mut state.nav, e),
        AppEvent::Notifications(e) => notifications::reduce(&mut state.notifications, e),
        AppEvent::Chat(e) => chat::reduce(&mut state.chat, e),
        AppEvent::Coop(e) => coop::reduce(&mut state.coop, e),
        AppEvent::Lobby(e) => lobby::reduce(&mut state.lobby, e),
        AppEvent::Replays(e) => replays::reduce(&mut state.replays, e),
        AppEvent::Maps(e) => maps::reduce(&mut state.maps, e),
        AppEvent::MapGenerator(e) => map_generator::reduce(&mut state.map_generator, e),
        AppEvent::Mods(e) => mods::reduce(&mut state.mods, e),
        AppEvent::Leaderboard(e) => leaderboard::reduce(&mut state.leaderboard, e),
        AppEvent::PlayerCard(e) => player_card::reduce(&mut state.player_card, e),
        AppEvent::Reporting(e) => reporting::reduce(&mut state.reporting, e),
        AppEvent::Reviews(e) => reviews::reduce(&mut state.reviews, e),
        AppEvent::Social(e) => social::reduce(&mut state.social, e),
        AppEvent::Tourney(e) => tourney::reduce(&mut state.tourney, e),
        AppEvent::Tutorials(e) => tutorials::reduce(&mut state.tutorials, e),
        AppEvent::Uploads(e) => uploads::reduce(&mut state.uploads, e),
        AppEvent::GalacticWar(e) => galactic_war::reduce(&mut state.galactic_war, e),
        AppEvent::ClientUpdate(e) => client_update::reduce(&mut state.client_update, e),
        AppEvent::Install(e) => install::reduce(&mut state.install, e),
        AppEvent::Settings(e) => settings::reduce(&mut state.settings, e),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::{ConnectionStatus, SessionEvent};

    #[test]
    fn routes_session_event_to_session_slice() {
        let mut state = AppState::default();
        reduce(
            &mut state,
            &SessionEvent::BackendReady {
                version: "9.9.9".into(),
                offline_auth: false,
            }
            .into(),
        );
        assert_eq!(state.session.status, ConnectionStatus::Connected);
        assert_eq!(state.session.backend_version, "9.9.9");
    }
}

File: ui/src/store/reducer.ts
// Frontend mirror of `faf-domain`'s reducer. Pure, immutable, and structurally
// identical to crates/faf-domain/src/reducer.rs: same events, same transitions.
// If you change a slice reducer in Rust, change its twin here (ARCHITECTURE.md §3.6).

import type { AppEvent, AppState } from "../ipc/bindings";
import { reduceChangelog } from "./reducers/changelog";
import { reduceChat } from "./reducers/chat";
import { reduceClientUpdate } from "./reducers/clientUpdate";
import { reduceCoop } from "./reducers/coop";
import { reduceAuth, reduceInstall, reduceNav, reduceSession, reduceSettings } from "./reducers/core";
import { reduceGalacticWar } from "./reducers/galacticWar";
import { reduceGuides } from "./reducers/guides";
import { reduceLeaderboard } from "./reducers/leaderboard";
import { reduceLobby } from "./reducers/lobby";
import { reduceMapGenerator } from "./reducers/mapGenerator";
import { reduceMaps } from "./reducers/maps";
import { reduceMods } from "./reducers/mods";
import { reduceNotifications } from "./reducers/notifications";
import { reducePlayerCard } from "./reducers/playerCard";
import { reduceReplays } from "./reducers/replays";
import { reduceReporting } from "./reducers/reporting";
import { reduceReviews } from "./reducers/reviews";
import { reduceSocial } from "./reducers/social";
import { reduceTraining } from "./reducers/training";
import { reduceTutorials } from "./reducers/tutorials";
import { reduceUploads } from "./reducers/uploads";
import { reduceTourney } from "./reducers/tourney";

export { isModerator, isPrivateChannel, mentions } from "./reducers/chat";
export { findPlayer, playersByNickname } from "./reducers/social";

export function applyEvent(state: AppState, event: AppEvent): AppState {
  switch (event.kind) {
    case "Session":
      return { ...state, session: reduceSession(state.session, event.event) };
    case "Auth":
      return { ...state, auth: reduceAuth(state.auth, event.event) };
    case "Nav":
      return { ...state, nav: reduceNav(state.nav, event.event) };
    case "Notifications":
      return { ...state, notifications: reduceNotifications(state.notifications, event.event) };
    case "Chat":
      return { ...state, chat: reduceChat(state.chat, event.event) };
    case "Changelog":
      return { ...state, changelog: reduceChangelog(state.changelog, event.event) };
    case "Coop":
      return { ...state, coop: reduceCoop(state.coop, event.event) };
    case "Install":
      return { ...state, install: reduceInstall(state.install, event.event) };
    case "Social":
      return { ...state, social: reduceSocial(state.social, event.event) };
    case "Lobby":
      return { ...state, lobby: reduceLobby(state.lobby, event.event) };
    case "Replays":
      return { ...state, replays: reduceReplays(state.replays, event.event) };
    case "Maps":
      return { ...state, maps: reduceMaps(state.maps, event.event) };
    case "MapGenerator":
      return { ...state, mapGenerator: reduceMapGenerator(state.mapGenerator, event.event) };
    case "Mods":
      return { ...state, mods: reduceMods(state.mods, event.event) };
    case "Leaderboard":
      return { ...state, leaderboard: reduceLeaderboard(state.leaderboard, event.event) };
    case "PlayerCard":
      return { ...state, playerCard: reducePlayerCard(state.playerCard, event.event) };
    case "Reporting":
      return { ...state, reporting: reduceReporting(state.reporting, event.event) };
    case "Reviews":
      return { ...state, reviews: reduceReviews(state.reviews, event.event) };
    case "Tourney":
      return { ...state, tourney: reduceTourney(state.tourney, event.event) };
    case "Guides":
      return { ...state, guides: reduceGuides(state.guides, event.event) };
    case "Training":
      return { ...state, training: reduceTraining(state.training, event.event) };
    case "Tutorials":
      return { ...state, tutorials: reduceTutorials(state.tutorials, event.event) };
    case "Uploads":
      return { ...state, uploads: reduceUploads(state.uploads, event.event) };
    case "GalacticWar":
      return { ...state, galacticWar: reduceGalacticWar(state.galacticWar, event.event) };
    case "ClientUpdate":
      return { ...state, clientUpdate: reduceClientUpdate(state.clientUpdate, event.event) };
    case "Settings":
      return { ...state, settings: reduceSettings(state.settings, event.event) };
  }
}
